src/Components/Components/src/Rendering/HtmlRenderer.cs

```diff
@@ -40,7 +43,13 @@ public HtmlRenderer(IServiceProvider serviceProvider, ILoggerFactory loggerFacto
         /// <inheritdoc />
         protected override Task UpdateDisplayAsync(in RenderBatch renderBatch)
         {
-            return Task.CompletedTask;
+            // By default we return a cancelled task. This has the effect of making it so that the
+            // OnAfterRenderAsync callbacks on components don't run by default.
+            // This way, by default prerendering gets the correct behavior and other renderers
+            // override the UpdateDisplayAsync method already, so those components can cancel
+            // either complete a task when the client acknowledges the render, or a cancelled task
+            // when the renderer gets disposed.
+            return CancelledRenderTask;
```

Comment:
TBH this feels quite hacky. I understand that it works, but it's relying on a pretty obscure behavior inside the `Renderer` base class, i.e., that if `UpdateDisplayAsync` returns a completed task, then it happens to skip the rest of the operations and doesn't raise any `OperationCanceledException` or similar. This is the sort of obscure coupling that could bite us in the future.

Rather than doing this, could the `Renderer` make use of the `IComponentContext` and contain its own logic to skip the "on after render" callbacks if the context says we're not connected? It can get the `IComponentContext` from the service provider in the constructor. It may be best to treat it as an optional dependency. I think we're at the point where `Renderer`'s set of DI dependencies isn't going to be statically defined, for forward-compatibility reasons.

Reply:
I could argue that the current behavior is wrong. It breaks the contract that OnAfterRender gets invoked when the batch has rendered on the target display. (In whatever form or shape that takes).

Given that the base class is not associated with any display technology and that the render never finishes, it makes sense to switch the implementation for a cancelled task.

I understand the "obscure coupling" argument, but I also believe that its not an obscure detail but front and center to the onafterrender contract. We've been telling developers that OnAfterRender runs when the display has been updated and references have been populated.

Our original default was wrong, because we would run it in a context where we can't actually perform a render, so returning Task.CompletedTask is bad.

My reasoning is that, Task.CancelledTask is the right thing to do because the base renderer will never complete an update and is only able to render static HTML. Derived classes provide the ability to update the display and complete a full render pass.

We also have quite a good amount of tests describing this behavior, so I'm not concerned about breaking anything without noticing.

Does ComponentContext.IsConnected change in circuits to signal disconnection? If so, I would stay away from using it for this, as we shouldn't simply avoid running the callbacks if we happen to hit !context.IsConnected

I do see your concerns but I think despite this being used to avoid running OnAfterRender callbacks, switching to Task.CancelledTask is the right thing to do for the base class, it just also happens to fix the problem.

Keep in mind that in RemoteRenderer we also cancel the completions when we dispose the RemoteRenderer, which is aligned with what this does here.

Reply:
OK, lots of good points there. I'm convinced by your argument that it's right to signal that we're not actually updating the display in the prerendering case. Just adding a comment to the place where we look for `IsCanceled` and saying this will be the case in prerendering would be enough to keep this from being too obscure in my opinion.

> Does ComponentContext.IsConnected change in circuits to signal disconnection? If so, I would stay away from using it for this, as we shouldn't simply avoid running the callbacks if we happen to hit !context.IsConnected

If I'm recalling correctly (please let me know if not), the `RemoteRenderer`'s `UpdateDisplay` returns a `Task` that only completes when we receive an ACK for that batch from the client. As such we only call `OnAfterRender` callbacks during periods when we're connected, not disconnected. This is good because people put JS interop code into `OnAfterRender` and they want its execution to be deferred until reconnection, otherwise it would fail.

I'm not sure that the `IsConnected` state can change during the synchronous period between receiving an ACK and when we invoke `OnAfterRender`.

So altogether I'm not sure there's any way we can hit !context.IsConnected. I'd be interested to know if you think we can.

Anyway this point is irrelevant if we're going ahead with the task cancellation, which I think I'm now happy with based on your reasoning. But I'd still like it if we amended [the comment here](https://github.com/aspnet/AspNetCore/blob/master/src/Components/Components/src/Rendering/Renderer.cs#L450) to be totally clear that this is a normal code path that happens every time during prerendering.